atlascope/core/job_types/nucleus_detection.py

```diff
@@ -68,25 +73,60 @@ def run(job_id: str, original_dataset_id: str):
     original_dataset = Dataset.objects.get(id=original_dataset_id)
     job = Job.objects.get(id=job_id)
 
+    def nucleus_data_subset(data: dict, subset_name: str):
+        return (
+            {
+                key.replace(subset_name, ''): value
+                for key, value in data.items()
+                if subset_name in key
+            },
+        )
+
     try:
         input_image = skimage.io.imread(
             io.BytesIO(original_dataset.content.read()),
         )
         nuclei = detect_nuclei(input_image)
 
-        job.resulting_datasets.add(
-            save_output_dataset(
-                original_dataset,
-                job.investigation,
-                'Detected Nuclei',
-                None,
-                {
-                    'num_nuclei': len(nuclei),
-                    'nucleus_detections': nuclei,
+        detection_dataset = save_output_dataset(
+            original_dataset,
+            job.investigation,
+            'Detected Nuclei',
+            None,
+            {
+                'num_nuclei': len(nuclei),
+            },
+            dataset_type='nucleus_detection',
+        )
+        for nucleus in nuclei:
+            DetectedNucleus.objects.create(
+                detection_dataset=detection_dataset,
+                label_integer=nucleus['Label'],
+                centroid=Point(
+                    x=nucleus['Identifier.CentroidX'],
+                    y=nucleus['Identifier.CentroidY'],
+                ),
+                weighted_centroid=Point(
+                    x=nucleus['Identifier.WeightedCentroidX'],
+                    y=nucleus['Identifier.WeightedCentroidY'],
+                ),
+                bounding_box=Polygon(
+                    [
+                        (nucleus['Identifier.Xmin'], nucleus['Identifier.Ymin']),
+                        (nucleus['Identifier.Xmax'], nucleus['Identifier.Ymin']),
+                        (nucleus['Identifier.Xmax'], nucleus['Identifier.Ymax']),
+                        (nucleus['Identifier.Xmin'], nucleus['Identifier.Ymax']),
+                        # repeat first point to create linear ring
+                        (nucleus['Identifier.Xmin'], nucleus['Identifier.Ymin']),
+                    ]
+                ),
+                **{
+                    nucleus_attribute_to_field_name(attribute): nucleus[attribute]
+                    for attribute in NUCLEUS_ATTRIBUTES
```

Comment:
This way of expressing it is a bit odd, and took me some reading of different parts of the code to understand what's going on. At the very least, I think you should factor this expression out to before the call to `create()` and construct the dict with an expressive name, then "dump" that object here with the `**` operator.

This is purely a readability thing, but that is an important consideration in complex code such as this.

Reply:
Ok, I changed this in [1394483](https://github.com/atlascope/atlascope/pull/169/commits/139448398be68ecada9512934a276c1ac7b11a16)